### ISSUE TITLE
Inconsistent treatment of Callable bound type vars between Callable args and function definition

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#2482]
**Describe the bug**
While I'm not too sure what's going on here, right now Pyright seems to be inconsistently passing over cases where Callable bound type-vars are present in the arguments of a separate Callable as part of a union (leading to an incompatibility error) while allowing a similar case in a function's definition.

**To Reproduce**
* Lint the example code.
* Watch it give the following error
```py
C:\Users\Lucina\PycharmProjects\tanjun\test.py
  C:\Users\Lucina\PycharmProjects\tanjun\test.py:30:2 - error: Argument of type "(ctx: int) -> Coroutine[Any, Any, None]" cannot be assigned to parameter of type "UsesFoo[CallbackSigT@also_takes_foo]"
    "object" is incompatible with "UsesFoo[CallbackSigT@also_takes_foo]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
Completed in 0.543sec
```

**Expected behavior**
Pyright should either complain about the usage of both `@takes_foo` and `@also_takes_foo` or more likely it should allow both cases (like MyPy currently does) and pass on the example code.

**Screenshots or Code**
```python
import typing

CallbackSig = typing.Callable[..., typing.Awaitable[None]]
CallbackSigT = typing.TypeVar("CallbackSigT", bound="CallbackSig")


class UsesFoo(typing.Generic[CallbackSigT]):
    def __init__(self, callback: CallbackSigT) -> None:
        self.callback = callback


FooIsh = typing.Union[CallbackSigT, UsesFoo[CallbackSigT]]


def takes_foo(foo: typing.Union[FooIsh[CallbackSigT]]) -> UsesFoo[CallbackSigT]:
    return UsesFoo(foo.callback if isinstance(foo, UsesFoo) else foo)


def also_takes_foo(
    b: int,
) -> typing.Callable[[FooIsh[CallbackSigT]], UsesFoo[CallbackSigT]]:
    return lambda foo: UsesFoo(foo.callback if isinstance(foo, UsesFoo) else foo)


@takes_foo
async def bar(ctx: int) -> None:
    ...


@also_takes_foo(123)
async def bars(ctx: int) -> None:
    ...

```

**VS Code extension or command-line**
Pyright as a command line tool: v1.1.181
